Added traits to facilitate instrumentation of Google APIs

ewiseblatt opened the pull request:
Instruments one operation as example of how it will be used.

@duftler 
This is independent of earlier PR, which I'm looking at now.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleExecutorTraits.groovy:
```diff
@@ -0,0 +1,86 @@
+     } catch (IOException ioex) {
+       registry.timer(id.withTag("success", "false")).record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS)
+       throw ioex;
+     } catch (Exception ex) {
```
If you are catching both IOException and Exception I suppose you could just catch Throwable instead.

ewiseblatt replied:
The IOException is handled different from the others. It is passed through. The caller might grab the HTTP response from HTTP-like exceptions. My thought here was that API calls are going to fail IO or HTTP exceptions (which are derived from IO) and get reported back. All the other exceptions and errors are more likely internal to Spinnaker implementation/configuration rather than user requests, resource state, and permissions.

duftler replied:
Sorry, that was a typo on my part. I intended to write: "If you are catching both `Exception` and `Error` I suppose you could just catch Throwable instead.".

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleExecutorTraits.groovy:
```diff
@@ -0,0 +1,86 @@
+     } catch (Exception ex) {
+       registry.timer(id.withTag("success", "false")).record(clock.monotonicTime() - startTime, TimeUnit.NANOSECONDS)
+       registry.counter(sizeId.withTag("success", "false")).increment(batch.size())
+       throw new IllegalStateException(ex);
```
Why IllegalStateException?

ewiseblatt replied:
I'm not sure. I'd seen that pattern elsewhere, but could have been in front50. I can take it out.

duftler replied:
If you've seen a similar pattern elsewhere in the neighborhood, I can live with it. Maybe `RuntimeException` would work too. I think you just want any unchecked exception here, and `RuntimeException` is the most general of those.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/ResizeGoogleServerGroupAtomicOperation.groovy:
```diff
@@ -63,17 +62,21 @@ class ResizeGoogleServerGroupAtomicOperation implements AtomicOperation<Void> {
-                                   serverGroupName,
-                                   targetSize).execute()
+      timeExecute(
+          instanceGroupManagers.resize(project,
```
I think the formatting might look better if it were indented only 2 spaces and the closing parens after `...TAG_REGION, region` were on the next line.
But I don't feel too strongly about it if you prefer it this way.

ewiseblatt commented:
FYI I changed things around to clean up and simplify as I mentioned off line yesterday.
The diff is "Refined and simplified" above. I'm going to submit since you already approved and saw the gist of the changes I planned on making.

duftler commented on clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/GoogleExecutorTraits.groovy:
```diff
@@ -0,0 +1,76 @@
+       batch.execute()
+       success = "true"
+     } finally {
+       def tagDetails = [TAG_BATCH_CONTEXT: batchContext, "success": success]
```
Pretty sure you'll need to use `[(TAG_BATCH_CONTEXT): batchContext...]` here.

```
def testKey = 'some-key'
def testMapBad = [testKey: 'some-value']
def testMapGood = [(testKey): 'some-value']
System.out.println("testMapBad=$testMapBad")
System.out.println("testMapGood=$testMapGood")
```

Yields:
```
testMapBad=[testKey:some-value]
testMapGood=[some-key:some-value]
```

duftler replied:
Also, the quotes can be dropped from any literal map keys like `"success"`.

ewiseblatt replied:
Interesting (though not in a good way). That also explains why I dont need the quotes on success. Thanks for catching that.

I've added in batching here, but I havent ported the batch operations so havent tested this.
